text/0000-unsafe-fields.md

```diff
@@ -0,0 +1,1037 @@
+
+Besides the unclear semantics of an unparameterized `unsafe()`, this design has the disadvantage
+that the most permissive (and thus dangerous) semantics are the cheapest to type. To mitigate this,
+we might instead imagine reversing the polarity of the modifier:
+
+- `safe(read)` all operations except reading are safe
+- `safe(read,&)` all operations except reading and `&`-referencing are safe
+- etc.
+
+...but using `safe` to denote the presence of a safety invariant is probably too surprising in the
+context of Rust's existing safety tooling.
+
+Alternatively, if we are confident that a hierarchy of operations exists, the brevity of the API can
+be improved by having the presence of one modifier imply others (e.g., `unsafe(&mut)` could denote
+that initialization, mutation and `&mut`-referencing) are unsafe. However, this requires that the
+user internalize this hierarchy, or else risk selecting the wrong modifier for their invariant.
+
+Although we cannot explore the entire design space of syntactic modifiers here, we broadly feel that
+their additional complexity exceeds that of our proposed design. Our proposed rule that *a field
+marked `unsafe` is unsafe to use* is both pedagogically simple and fail safe; i.e., so long as a
+field is marked `unsafe`, it cannot be misused in such a way that its invariant is violated in safe
+code.
+
+### Mixing Syntactic Knobs with a Wrapper Type
+
+One alternative proposed in this RFC's discussion recommends a combination of syntactic knobs and a
+wrapper type. In brief, a simple [`Unsafe` wrapper type](#unsafe-wrapper-type) would be provided,
+along with two field safety modifiers:
+
+- `unsafe`
+  All uses except reading are `unsafe`.
+- `unsafe(mut)`
+  All uses except reading and `&`-referencing are `unsafe`.
+
+Under this proposal, a programmer would use some combination of `unsafe`, `unsafe(mut)` and `Unsafe`
+to precisely tune Rust's safety tooling protections, depending on the hazards of their invariant.
+
+The primary advantage of this approach is that it results in comparatively fewer instances in which
+[the programmer must write a 'trivial' safety proof](#trivial-safety-proofs). However, it achieves
+this by front-loading the requirement that the programmer imagine all possible safety hazards of
+their field. A mistake, here, may lead to a false sense of security if Rust fails to require
+`unsafe` for uses that are, in fact, dangerous. By contrast, this RFC requires that programmers
+resolve these questions only on an as-needed basis; e.g., until you need to `&`-reference a field,
+you do not need to confront whether doing so is *always* a safe operation.
+
+This alternative also inherits some of the disadvantages of [`Unsafe` wrapper
+types](#unsafe-wrapper-type); namely that the safety proofs needed to operate on an `Unsafe` wrapper
+value depend on the dataflow of the program; the wrapper value must be traced to its originating
+field so that field's safety documentation may be examined.
```

Comment:
Again, the same is true of this RFC’s design. With this RFC, the `unsafe` is required at the site of the copy/move, and to determine whether the safety condition is met, you must trace dataflow forward to where the value is later used. With `Unsafe` wrappers, the `unsafe` is required at the use-site, and dataflow must be traced back to the copy/move site. Both solutions require tracing dataflow, just in different directions.

Reply:
But isn't that doubly true for `Unsafe`? `Unsafe`'s accessors also require unsafe, forcing you to further trace the dataflow forward to assess that uses don't violate the field invariant (which you discovered by tracing the dataflow backwards).